### PR TITLE
feat: introduce parsing for lengths

### PR DESCRIPTION
**Pull request by CheesyGamer77**
This PR introduces the ability to parse lengths and their respective units, including:

- [x] Units (`<length-unit>`)
- [x] Length (`<length>`)
- [x] Percentages (`<percentage>`)
- [x] Length or Percentage (`<length-or-percentage>`)
- [x] Side-Lengths (`<side-length>`)
  - [x] Single
  - [x] Double
  - [x] Quad

### CODE REVIEW

File: src/parser/length/tests.rs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+mod units {
+    use super::*;
+
+    #[test]
+    fn nothing() {
+        let input = "";
+
+        assert!(DistanceUnit::from_str(input).is_err());
```
Comment (nstringham) [inline src/parser/length/tests.rs:15]: You could also use .parse if you want

File: src/parser/length.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    Pixels,          // px
+    Centimeters,     // cm
+    Inches,          // in
+    Points,          // pt
+    FontSize,        // em
+    RootFontSize,    // rem
+    ViewportHeight,  // vh
+    ViewportWidth,   // vw
+    RootBlockSize,   // vb
+    RootInlineSize,  // vi
```
Comment (nstringham) [inline src/parser/length.rs:27]: Should probably start with "Viewpoint" for consistentcy

File: src/parser/length.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+use crate::tokenizer::*;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug, PartialEq)]
+pub enum Distance {
+    Zero(),
+    Distance(f64, DistanceUnit),
+    Percentage(f64),
```
Comment (nstringham) [inline src/parser/length.rs:13]: I don't think be a percentages should be a valid length.

File: src/parser/length.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+impl Parsable for Distance {
+    fn parse<I: Iterator<Item = char>>(parser: &mut Parser<I>) -> Result<Self, ParsingError> {
+        match parser.tokens.next() {
+            Some(token_at) => match token_at.token {
+                Token::Number(value) if value == 0.0 => Ok(Distance::Zero()),
+                Token::Percentage(value) if value >= 0.0 => Ok(Distance::Percentage(value)),
+                Token::Dimension(value, unit) => Ok(Distance::Distance(
+                    value,
+                    DistanceUnit::from_str(unit.as_str()).unwrap(),
+                )),
```
Comment (nstringham) [inline src/parser/length.rs:65]: Please no unwrap in production code. You should add a test for a number with an invalid unit to catch this case

**Review by nstringham — CHANGES_REQUESTED** [review]

File: src/parser/length.rs (hunk 2 above)
Comment (CheesyGamer77) [reply src/parser/length.rs:27]: So like `ViewpointRootBlockSize` and `ViewpointRootInlineSize` correct?

Comment (nstringham) [reply src/parser/length.rs:27]: I don't think you need "root" really

**Comment by nstringham** [comment]
for future reference it would probably have bean better to do percentages on a separate branch so that length could be merged

**Comment by nstringham** [comment]
Please run
```
cargo fmt
```

**Comment by nstringham** [comment]
Clippy thinks you have a redundant `return` somewhere in length

**Comment by nstringham** [comment]
It seems strange to have LenghtOrPercentage in the same file as length when Percentage gets it's own file

File: src/parser/length_or_percentage.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    Percentage(Percentage),
+}
+
+impl Parsable for LengthOrPercentage {
+    fn parse<I: Iterator<Item = char>>(parser: &mut Parser<I>) -> Result<Self, ParsingError> {
+        match parser.tokens.peek() {
+            Some(token_at) => match token_at.token {
+                // parse length
+                Token::Number(_) | Token::Dimension(_, _) => {
+                    Ok(LengthOrPercentage::Length(parser.parse::<Length>()?))
```
Comment (nstringham) [inline src/parser/length_or_percentage.rs:17]: `::<Length>` is optional here

File: src/parser/length_or_percentage.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                }
+
+                // parse percentage
+                Token::Percentage(_) => Ok(LengthOrPercentage::Percentage(
+                    parser.parse::<Percentage>()?,
```
Comment (nstringham) [inline src/parser/length_or_percentage.rs:22]: `::<Percentage>` is optional here

File: src/parser/rule/ruleset.rs [hunk 7]
```diff
@@ -4,7 +4,7 @@ use super::*;
 pub struct Ruleset {}
 
 impl Parsable for Ruleset {
-    fn parse<I: Iterator<Item = char>>(parser: &mut Parser<I>) -> Result<Self, ParsingError> {
+    fn parse<I: Iterator<Item = char>>(_: &mut Parser<I>) -> Result<Self, ParsingError> {
```
Comment (nstringham) [inline src/parser/rule/ruleset.rs:7]: do we want to do this here or keep it until we implement this?

File: src/parser/side.rs [hunk 8]
```diff
@@ -47,6 +47,8 @@ pub trait CanStart: Parsable {
 
 #[cfg(test)]
 mod tests {
+    use crate::parser::{length::Length, length_or_percentage::LengthOrPercentage};
+
     use super::{color::Color, *};
```
Comment (nstringham) [inline src/parser/side.rs:52]: `crate::parser` is the same as `super` we should pick one for consistancy

File: src/parser/percentage.rs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            None => Err(ParsingError::end_of_file("percentage")),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn parse_percentage(input: &str) -> Result<Percentage, ParsingError> {
```
Comment (nstringham) [inline src/parser/percentage.rs:24]: I think you only use this function once

File: src/parser/length/tests.rs [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+            assert_eq!(parse_unit("vmax"), Ok(LengthUnit::ViewportMaximum));
+        }
+
+        #[test]
+        fn invalid_unit() {
+            assert!(parse_unit("xd").is_err());
+        }
+    }
+
+    mod length_or_percentage {
```
Comment (nstringham) [inline src/parser/length/tests.rs:91]: why is this in length?

**Review by nstringham — CHANGES_REQUESTED** [review]

File: src/parser/rule/ruleset.rs (hunk 7 above)
Comment (CheesyGamer77) [reply src/parser/rule/ruleset.rs:7]: I don't mind either way personally. I was just going to leave this alone tbh

File: src/parser/length/tests.rs (hunk 10 above)
Comment (CheesyGamer77) [reply src/parser/length/tests.rs:91]: Forgot to move that over, whoops

File: src/parser/rule/ruleset.rs (hunk 7 above)
Comment (nstringham) [reply src/parser/rule/ruleset.rs:7]: Personally if I would leave it `parser` but if the clippy errors get to annoying then we should do `_parser` not `_` IMO

**Review by nstringham — APPROVED** [review]

**Pull request merged**